xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.fromThrowable(t).augmentDescription("ADS stream [" + this + "] had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("ADS stream [" + this + "] was closed"));
+        }
+      });
+    }
+
+    private void handleStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      configWatcher.onError(error);
+      if (responseReceived || retryBackoffPolicy == null) {
+        // Reset the backoff sequence if had received a response, or backoff sequence
+        // has never been initialized.
+        retryBackoffPolicy = backoffPolicyProvider.get();
+      }
+      long delayNanos = 0;
+      if (!responseReceived) {
+        delayNanos =
+            Math.max(
+                0,
+                retryBackoffPolicy.nextBackoffNanos() - stopwatch.elapsed(TimeUnit.NANOSECONDS));
+      }
+      logger.log(Level.FINE, "{0} stream closed, retry in {1} ns", new Object[]{this, delayNanos});
+      if (delayNanos == 0) {
+        startRpcStream();
+        if (configWatcher != null) {
+          adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
+        }
+        // TODO(chengyuanzhang): send CDS/EDS requests if CDS/EDS watcher presents.
+      } else {
+        rpcRetryTimer =
+            syncContext.schedule(
+                new RpcRetryTask(), delayNanos, TimeUnit.NANOSECONDS, timeService);
+      }
+    }
+
+    private void close(Exception error) {
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      requestWriter.onError(error);
+    }
+
+    private void cleanUp() {
+      if (adsStream == this) {
+        adsStream = null;
+      }
+    }
+
+    /**
+     * Sends a DiscoveryRequest for the given resource name to management server. Memories the
+     * requested resource name (except for LDS as we always request for the singleton Listener)
+     * as we need it to find resources in responses.
+     */
+    private void sendXdsRequest(String typeUrl, Collection<String> resourceNames) {
+      checkState(requestWriter != null, "ADS stream has not been started");
+      String version = "";
+      String nonce = "";
+      if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+        version = ldsVersion;
+        nonce = ldsRespNonce;
+      } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+        checkArgument(resourceNames.size() == 1,
+            "RDS request requesting for more than one resource");
```

Comment:
Here is the background story:

Originally we only need `private void sendXdsRequest(String typeUrl, String resourceName) {...}` as for both LDS and RDS, we would set at most one `resource_name`. But for CDS/EDS to be added later, we need `private void sendXdsRequest(String typeUrl, Collection<String> resourceNames) {...}` as we may need to set more than one `resource_names` in requests. Since this is private and internal, we do not bother create method overloading. So just unify them into `private void sendXdsRequest(String typeUrl, Collection<String> resourceNames) {...}` and we send LDS/RDS requests through this method as well. So this check a good-to-have check, not strictly necessary.

The field `AdsStream.rdsResourceName`, as its comment says, it memorizes the resource name in the most recently sent RDS request. For LDS, we don't need such a memorization as it never changes.